tests/test_2408_layoutbuilder_in_numba.py

```diff
@@ -0,0 +1,1508 @@
+            lb.ListOffset(np.int64, lb.Numpy(np.int32)),
+        ],
+    )
+    assert len(builder) == 0
+
+    # error = ""
+    # assert builder.is_valid(error) == True
+
+    one = builder.append_content(0)
+    one.append(1.1)
+
+    # assert builder.is_valid(error) == True
+
+    two = builder.append_content(1)
+    list = two.begin_list()
+    list.append(1)
+    list.append(2)
+    two.end_list()
+
+    # assert builder.is_valid(error) == True
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [1.1, [1, 2]]
+
+
+def test_Union_ListOffset_Record():
+    builder = lb.Union(
+        np.int8,
+        np.int64,
+        [
+            lb.ListOffset(np.int64, lb.Numpy(np.int32)),
+            lb.Record([lb.Numpy(np.float64), lb.Numpy(np.int64)], ["x", "y"]),
+        ],
+    )
+    assert len(builder) == 0
+
+    one = builder.append_content(0)
+    list = one.begin_list()
+    list.append(1)
+    list.append(3)
+    one.end_list()
+
+    two = builder.append_content(1)
+    x = two.field("x")
+    y = two.field("y")
+
+    x.append(1.1)
+    y.append(11)
+
+    builder.append_content(0)
+    list = one.begin_list()
+    list.append(5.5)
+    one.end_list()
+
+    builder.append_content(1)
+    x.append(2.2)
+    y.append(22)
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [[1, 3], {"x": 1.1, "y": 11}, [5], {"x": 2.2, "y": 22}]
+
+
+def test_unbox():
+    @numba.njit(debug=True)
+    def f1(x):
+        x  # noqa: B018 (we want to test the unboxing)
+        return 3.14
+
+    builder = lb.BitMasked(np.uint8, lb.Numpy(np.float64), True, True)
+    f1(builder)
+
+    builder = lb.ByteMasked(bool, lb.Numpy(np.float64), valid_when=True)
+    f1(builder)
+
+    builder = lb.Empty()
+    f1(builder)
+
+    builder = lb.IndexedOption(np.int64, lb.Numpy(np.float64))
+    f1(builder)
+
+    builder = lb.IndexedOption(np.int32, lb.Numpy(np.float64))
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.ListOffset(np.int64, lb.Numpy(np.int64)))
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.Numpy(np.float64))
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.Empty())
+    f1(builder)
+
+    builder = lb.ListOffset(np.int8, lb.Numpy(np.int64))
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.Empty())
+    f1(builder)
+
+    builder = lb.Numpy(np.int32)
+    f1(builder)
+
+    builder = lb.Record(
+        [
+            lb.Numpy(np.float64),
+            lb.Numpy(np.int64),
+            lb.Numpy(np.uint8),
+        ],
+        ["one", "two", "three"],
+    )
+    f1(builder)
+
+    @numba.njit(debug=True)
+    def f_compiled_index(x):
+        i = x._contents[0]
+        i = x._contents[1]
+        i = x._contents[2]
+        return i
+
+    f_compiled_index(builder)
+
+    @numba.njit(debug=True)
+    def f_runtime_index(x, indx):
+        i = x._contents[indx]
+        return i
+
+    # FIXME: Tuple needs a compile-time index
+    # f_runtime_index(builder, 0)
```

Comment:
Since each field of a Tuple can have a different type, indexing can't be runtime.

```python
    @numba.njit
    def f_runtime_index(x, indx):
        i = x._contents[indx]
        return i
```

_should_ fail to compile because if `indx == 0`, the return type of the function would be one thing, and if `indx == 1`, the return type of the function would be another thing, etc.

Just as Record requires a literal string to select a field, Tuple requires a literal integer to select a field.